connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala

```diff
@@ -1144,32 +1154,87 @@ abstract class AvroSuite
     }
   }
 
-  test("unsupported nullable avro type") {
+  test("int/long double/float conversion") {
     val catalystSchema =
       StructType(Seq(
-        StructField("Age", IntegerType, nullable = false),
-        StructField("Name", StringType, nullable = false)))
+        StructField("Age", LongType),
+        StructField("Length", DoubleType),
+        StructField("Name", StringType)))
 
-    for (unsupportedAvroType <- Seq("""["null", "int", "long"]""", """["int", "long"]""")) {
+    for (optionalNull <- Seq(""""null",""", "")) {
       val avroSchema = s"""
         |{
         |  "type" : "record",
         |  "name" : "test_schema",
         |  "fields" : [
-        |    {"name": "Age", "type": $unsupportedAvroType},
+        |    {"name": "Age", "type": [$optionalNull "int", "long"]},
+        |    {"name": "Length", "type": [$optionalNull "float", "double"]},
         |    {"name": "Name", "type": ["null", "string"]}
         |  ]
         |}
       """.stripMargin
 
       val df = spark.createDataFrame(
-        spark.sparkContext.parallelize(Seq(Row(2, "Aurora"))), catalystSchema)
+        spark.sparkContext.parallelize(Seq(Row(2L, 1.8D, "Aurora"))), catalystSchema)
 
       withTempPath { tempDir =>
-        val message = intercept[SparkException] {
+        df.write.format("avro").option("avroSchema", avroSchema).save(tempDir.getPath)
+        checkAnswer(
+          spark.read
+            .format("avro")
+            .option("avroSchema", avroSchema)
+            .load(tempDir.getPath),
+          df)
+      }
+    }
+  }
+
+  test("non-matching complex union types") {
+    val catalystSchema =
+      StructType(Seq(
+        StructField("Union", StructType(Seq(
+          StructField("member0", IntegerType),
+          StructField("member1", StructType(Seq(StructField("f1", StringType, nullable = false))))
+        )))))
+
+    val df = spark.createDataFrame(
+      spark.sparkContext.parallelize(Seq(Row(Row(1, null)))), catalystSchema)
+
+    val recordSchema = """{"type":"record","name":"r","fields":[{"name":"f1","type":"string"}]}"""
+    for ((unionSchema, compatible) <- Seq(
+      (""""null","int",""" + recordSchema, true),
+      (""""int","null",""" + recordSchema, true),
+      (""""int",""" + recordSchema + ""","null"""", true),
+      (""""int",""" + recordSchema, true),
+      (""""null",""" + recordSchema + ""","int"""", false),
+      (""""null",""" + recordSchema, false),
+      (""""null","int",{"type":"record","name":"r","fields":[{"name":"f2","type":"string"}]}""",
+        false)
+    )) {
+      val avroSchema = s"""
+                          |{
+                          |  "type" : "record",
+                          |  "name" : "test_schema",
+                          |  "fields" : [
+                          |    {"name": "Union", "type": [$unionSchema]}
+                          |  ]
+                          |}
+      """.stripMargin
```

Comment:
Maybe use `SchemaBuilder` here? It was hard for me to read the unionSchema examples with so many quotes
```suggestion
    val recordS = SchemaBuilder.record("r").fields().requiredString("f1").endRecord()
    val intS = Schema.create(Schema.Type.INT)
    val nullS = Schema.create(Schema.Type.NULL)
    for ((unionTypes, compatible) <- Seq(
      (Seq(nullS, intS, recordS), true),
      (Seq(intS, nullS, recordS), true),
      (Seq(intS, recordS, nullS), true),
      (Seq(intS, recordS), true),
      (Seq(nullS, recordS, intS), false),
      (Seq(nullS, recordS), false),
      (Seq(nullS, SchemaBuilder.record("r").fields().requiredString("f2").endRecord()), false)
    )) {
      val avroSchema = SchemaBuilder.record("test_schema").fields()
        .name("union").`type`(Schema.createUnion(unionTypes: _*)).noDefault()
        .endRecord().toString()
```